### DISCUSSION TITLE
Need explanation of bitSizeOf(), initializeOffsets(), and hashCode() functions to test our C++ code

### DISCUSSION THREAD

**Discussion by bhanuprakasheagala-at-mbrdi** [ndsev/zserio#666] [Q&A]
Hi Team,

 We are currently working on writing unit tests for three functions in zserio generated **C++ code** for our project, but having difficulty understanding how they work and what to expect as final value comparisons. I would greatly appreciate it if you could provide some clarification on the following:

**bitSizeOf() :**
As per the reference document, it's mentioned as it calculates a number of bits needed for serialization of the Zserio object. Need few clarifications:
Do we need to manually calculate the total number of bits based on the fields defined in the structure, if so is there any alignment or padding needs to be added?
Are there any edge cases or special scenarios I should consider when testing?

**initializeOffsets() :**
From the name of the function, it seems like it sets offsets for fields. But it would be very helpful if there is some more information on how it works, what kind of inputs and outputs, and the expected behavior.

**hashCode() :**
As per the README reference, it calculates a hash code of the Zserio object. Could you please help us with more information on how it calculates, how it works, and the expected outputs.

In summary, In order to test our code, we need to understand how these functions are working/calculating the required values. It will help us to write EXPECT_EQ/ASSERT_EQ statements with a specific value to compare against the calculated values.

Please reach out if any more clarifications needed from our end.

Thankyou

**Comment by mikir** [comment]
Hello,

welcome and thanks for the question. I will try to answer:

**bitSizeOf()**:
You might try to calculate the total number of bits bases on the fields but it might be difficult in case of complex structures. The calculation is not straightforward, you need to understand Zserio encoding rules in detail. These encoding rules are described in the [Zserio Encoding Guide](https://zserio.org/doc/ZserioEncodingGuide.html).

I recommend to test `bitSizeOf()` method by using the `BitStreamWriter.getBitPosition()` method. You can write the data to the stream, get the current bit position after write and compare it to the bit size:

```
zserio::BitBuffer bitBuffer(zserioObject.bitSizeof());
zserio::BitStreamWriter writer(bitBuffer);
zserioObject.write(writer);
ASSERT_EQ(zserioObject.bitSizeOf(), writer.getBitPosition());
```

**initializeOffsets()**:
This is Zserio internal method. Not sure if you must test Zserio internal methods as well. It initializes offsets which are defined in the schema. Offsets are described [here](https://zserio.org/doc/ZserioLanguageOverview.html#offsets). If you really need to check this, I would recommend to use the similar approach as for `bitSizeOf()`. If you need the description of this method, inputs and outputs, you might use the command line option `-withCodeComments` during the generation. This option enforces generation of the comprehensive comments in the code.

**hashCode()**:
The hash code calculation is implemented in the runtime module [`HashCodeUtil.h`](https://github.com/ndsev/zserio/blob/master/compiler/extensions/cpp/runtime/src/zserio/HashCodeUtil.h). However, if you need to test it, you will not probably want to calculate it again with the same algorithm. It is difficult to make any advice here. I would suggest to calculate the hash code once and put to the assert the result:

```
// use hardcoded values to check that the hash code is stable
ASSERT_EQ(1178167, zserioObject.hashCode());
```

I hope, it helps. Don't hesitate to ask again if something is still not clear.
